spec/consensus/light-client.md

```diff
@@ -316,3 +316,70 @@ func Backwards(h1,h2) bool {
   return (hash(h2) == old.Header.hash)
  }
 ```
+
+### Failure Detection Module
+
+The light client must be able to detect failures in the trust assumptions.
+Ideally, it should never trust a header, which was forged by malicious
+validators.
+
+### Connectivity requirement
+
+To meet the above requirement, it must be connected to at least one honest
+full node. This is sufficient so long as there's no fork on the main chain.
+
+If there is a fork on the main chain, it means that two full nodes have decided
+on two different headers for the same height. Then the requirement would at
+least require that the lite client is **connected to one honest full node on
+each branch of the fork**.
+
+_Assumption_: no more than `2f` of the nodes are malicious (`f+1` are honest).
+otherwise, they could create more forks.
+
+In practice, this means connecting to 1 or more geographically distributed full
+nodes (called witnesses), which belong to different companies. Note this number
+does not include the primary full node (called primary), which is used for
+obtaining new headers.
+
+_Remark_: we can't guarantee all witnesses won't follow the same branch of a
+fork. to guarantee that, we'll need the light client to be connected to +1/3 of
+nodes, which is impractical.
+
+### Evidence detection
+
+After the light client verifies a new header (`H`) it received from primary, it
+should cross-check `H` with the headers from all witnesses. Cross-checking
+means comparing hashes of the headers. If any two hashes (or more) diverge,
+there's a fork (on the main chain OR phantom fork targeting this light client).
+
+The light client will then need to validate the header it got from a witness
+(`H1`) and verify it's signed by at least one trusted validator. If either of
+these fails, it should disconnect from the offending witness.
+
+1. Equivocation
+
+  If some validator double signed, the light client should form & submit a
+  `DuplicateVoteEvidence`.
+
+  After doing so, the light client must stop its operation. The operator will be
+  forced to reset the light client (resetting does not imply deleting the data
+  here) with a new trusted header.
+
```

Comment:
I guess we need to be precise about what data can be maintained and how the detector interfaces with the bisection. If the lite client trusts a header h' and has used h during bisection to get this trust, and then later it turns out h is part of a fork, then the trust in h' needs to be revoked also. If h' has already been used, I guess we might have to reset in this case.

Reply:
> the trust in h' needs to be revoked also

ugh... this won't make retail shops happy ("here's your coffee John. enjoy! wtf, wait! looks like your transaction only was committed in a fork")

Reply:
In any case if we detect a fork the light client needs to halt and be manually re-initialized.